feat: Organization VPCs and org VPC peering (#579)
aiven/aiven-docs: docs/platform/howto/_manage-org-vpc-peering-azure.md

---
title: Manage an organization VPC peering with Microsoft Azure
sidebar_label: Azure peering
---

import DeleteOrgPeering from "@site/static/includes/vpc/delete-org-peering.md";
import ConsoleLabel from "@site/src/components/non-swizzled/ConsoleIcons";
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

Set up a peering connection between your [Aiven organization VPC](/docs/platform/howto/manage-organization-vpc) and a [Microsoft Azure virtual network](https://learn.microsoft.com/en-us/azure/virtual-network/create-peering-different-subscriptions).

Establishing a peering connection between an Aiven organization VPC and an Azure VNet
requires creating the peering both from the VPC in Aiven and from the VNet in Azure.

To establish the peering from Aiven to Azure, the Aiven Platform's
[Active Directory application object](https://learn.microsoft.com/en-us/azure/active-directory/develop/app-objects-and-service-principals)

> needs permissions in your Azure subscription. Because the peering is between different AD
> tenants (the Aiven AD tenant and your Azure AD tenant), another application object is
> needed for your Azure AD tenant to create the peering from Azure to Aiven once granted
> permissions to do so.

Reviewer: ```suggestion
needs permissions in your Azure subscription. Because the peering is between different AD 
tenants (the Aiven AD tenant and your Azure AD tenant), your Azure AD tenant needs another 
application object. Once granted permissions, this object allows peering 
from Azure to Aiven.
```

how about breaking the line for added clarity?

  optionally, the [Microsoft Azure portal](https://portal.azure.com/#home)
- Access to the [Aiven Console](https://console.aiven.io/)
- [Aiven CLI](/docs/tools/cli) installed
- Two VPCs to be peered: an
  [organization VPC](/docs/platform/howto/manage-organization-vpc#create-an-organization-vpc)
  in Aiven and a VNet in your Azure account

## Set up permissions in Azure

### Azure app object permissions

1. Log in with an Azure admin account using the Azure CLI:

   ```bash
   az account clear
   az login
   ```

   This should open a window in your browser prompting to choose an Azure
   account to log in with.

   :::tip
   If you manage multiple Azure subscriptions, also configure the Azure CLI
   to default to the correct subscription for the subsequent commands. This
   is not needed if there's only one subscription:

   ```bash
   az account set --subscription SUBSCRIPTION_NAME_OR_ID
   ```

   :::

1. Create an application object in your AD tenant using the Azure CLI:

   ```bash
   az ad app create                         \
     --display-name "NAME_OF_YOUR_CHOICE"   \
     --sign-in-audience AzureADMultipleOrgs \
     --key-type Password
   ```

>    This creates an entity to your AD that can be used to log into multiple

Reviewer: ```suggestion
   This creates an application object in Azure AD that can be used to log into multiple
```

   AD tenants ( `--sign-in-audience AzureADMultipleOrgs` ), but only the
   home tenant (the tenant the app was created in) has the credentials to
   authenticate the app.

   :::note
   Save the `appId` field from the output. It will be referred to as `$user_app_id`.
   :::

> 1. Create a service principal for your app object to the Azure subscription that the VNet
>    to be peered is located in:

Reviewer: ```suggestion
1. Create a service principal for your app object in the Azure subscription where the VNet 
   to be peered is located in: 
```
to be more precise?

   ```bash
   az ad sp create --id $user_app_id
   ```

>    This creates a service principal to your subscription that may be given
>    permissions to peer your VNet.

Reviewer: ```suggestion
   This creates a service principal in your subscription, which can be 
   assigned permissions to peer your VNet.
```
Maybe something like this? I feel the preposition "to" in this context is incorrect.

   - Azure Subscription ID (the VNet page in the
     [Azure portal](https://portal.azure.com/#home) > **Essentials** > **Subscription ID**)
     or the part after `/subscriptions/` in the resource ID. It will be referred to as
     `$user_subscription_id`.
   - Resource group name (the VNet page in the [Azure portal](https://portal.azure.com/#home) >
     **Essentials** > **Resource group**) or the `resourceGroup` field in the output. This
     will be referred to as `$user_resource_group`.
   - VNet name (title of the VNet page), or the `name` field from
     the output. It will be referred to as `$user_vnet_name`.

   :::note
   Save all the properties for later.
   :::

1. Grant your service principal permissions to peer.

   The service principal needs to be assigned a role that has the permission for the
   `Microsoft.Network/virtualNetworks/virtualNetworkPeerings/write` action
   on the scope of your VNet. To limit the amount of permissions the application
   object and the service principal have, you can create a custom role with just

>    that permission. The built-in network contributor role includes that permission.

Reviewer: How about something like this here: 

The service principal needs to be assigned a role that includes the `Microsoft.Network/virtualNetworks/virtualNetworkPeerings/write` permission at the scope of your VNet. To limit the permissions granted to the application object and the service principal, you can create a custom role with only this permission. The built-in Network Contributor role also includes this permission.

   1. Find the id of the role with the required permission:

      ```bash
      az role definition list --name "Network Contributor"
      ```

>       The `id` field from the output will be referred to as `$network_contributor_role_id`.

Reviewer: ```suggestion
      The `id` field in the output is referred to as `$network_contributor_role_id`.
```

   1. Assign the service principal the network contributor role using
      `$network_contributor_role_id`:

      ```bash
      az role assignment create             \
        --role $network_contributor_role_id \
        --assignee-object-id $user_sp_id    \
        --scope $user_vnet_id
      ```

      This allows your application object to manage the network in the `--scope`.
      Since you control the application object, it may also be given permission for the
      scope of an entire resource group or the whole subscription to allow creating other

>       peerings later without assigning the role for each VNet separately.

Reviewer: For added clarity: 

This allows your application object to manage the network within the specified `--scope`. Since you control the application object, you can also grant it permissions at the scope of an entire resource group or the whole subscription. This enables creating other peerings later without assigning the role to each VNet separately.

   1. Click **Create**.

      This adds a connection with the **Pending peer** status in the
      [Aiven Console](https://console.aiven.io/).
1. While still on the **VPC details** page, make a note of the **ID** of your Aiven VPC.

</TabItem>
</Tabs>

-->

## Create the peering in Azure

[Establish the peering connection](https://learn.microsoft.com/en-us/azure/virtual-network/create-peering-different-subscriptions)
from your Azure VNet to the Aiven organization VPC:

1. Log out the Azure user
   [you logged in with](/docs/platform/howto/vnet-peering-azure#azure-app-object-permissions):

   ```bash
   az account clear
   ```

1. Log in the Azure application object to your AD tenant using the
   [password](/docs/platform/howto/vnet-peering-azure#azure-app-object-permissions):

   ```bash
   az login              \
     --service-principal \
     -u $user_app_id     \
     -p $user_app_secret \
     --tenant $user_tenant_id
   ```

1. Log in the Azure application object to the Aiven AD tenant:

   ```bash
   az login              \
     --service-principal \
     -u $user_app_id     \
     -p $user_app_secret \
     --tenant $aiven_tenant_id
   ```

>    At this point, your application object should have an opened session with your Azure AD

Reviewer: ```suggestion
   At this point, your application object should have an open session with your Azure AD
```
